### PR TITLE
Variables in OSL [WIP]

### PR DESCRIPTION
**Pull request by rgrinberg**
This is the first version that sketches the support. The aim is to get something working first in this PR. Do a subsequent where `:include` is removed in dune files (with or without an appropriate warning?), and then add properly optimized implementation if necessary.

@diml I'd like some early review on this as some things aren't turning as clean as I'd like them to be. Like constructing that map for read and read-lines variables and having yet another expansion function in that file.

### CODE REVIEW

**Comment by diml** [comment]
It would indeed be better if the code expanding the special forms could be shared between actions and sets, otherwise it's going to be hard to make sure they are in sync.

BTW, I have been thinking about this again and I think we shouldn't make this a pre-requisite for 1.0.0. We should be able to add support for `%{read:...}` and other special forms after 1.0.0 without breaking things given that we now reject all unknown variables in dune files. This only means that we'll have to support `:include` until 2.0.0, but that's not too bad.

**Comment by rgrinberg** [comment]
Okay, so implementing this PR without incurring any extra boilerplate should be quite possible.

**Comment by rgrinberg** [comment]
@diml how permissive should we be when expanding OSL forms btw? Allowing all of them seems to be the easiest. I.e. stuff like expanding `%{bin:foo.exe}` and adding `foo.exe` as a dep where the OSL is expanded.

Note that this is more than what we orignally planned: only allowing `%{read-xxx:..}` and friends.

**Comment by diml** [comment]
Allowing everything seems best

**Comment by rgrinberg** [comment]
Pushing this along. Another round of review from @diml and @emillon would be welcome. The main change is that we require a 2 stage expansion process for OSL inspired by how we do it for actions. It would be nice to short circuit this approach when we manage to expand all variables at the first stage but I'm not sure what's a clean way to accomplish that. There's also some still duplicate code between `Resolved_forms` and `expand_step*` family of functions.

The diff looks so big because a bunch of stuff had to be rearranged in super_context, but it's not that scary.

I'd like to move this along so that we can start encouraging users to move to this syntax ASAP, and because I've already spent far too much time on this little feature :)

File: src/ordered_set_lang.ml [hunk 1]
```diff
@@ -297,48 +336,46 @@ module Unexpanded = struct
...
-        let open Stanza.Of_sexp in
-        parse
-          (Parse.without_include ~elt:(String_with_vars.t >>| f_elems))
-          context
-          sexp
+      | Include (Template_path fn) ->
+        begin match f.f ~mode:Single fn with
+        | Unexpanded e ->
+          Loc.fail (String_with_vars.loc fn)
+            "Failed to fully expand include path. Unexpanded value: %a"
```
Comment (diml) [inline src/ordered_set_lang.ml]: I'm not sure this is very clear for users. It's more of an implementation detail. What about pointing to the first percent form in `e` and saying: `This percent form is not supported in this position`?

File: src/ordered_set_lang.mli [hunk 2]
```diff
@@ -62,21 +74,17 @@ module Unexpanded : sig
...
-      filenames to their parsed contents. Every [(:include fn)] in [t] is replaced by
-      [Map.find files_contents fn]. Every element is converted to a string using [f]. *)
   val expand
     :  t
     -> dir:Path.t
-    -> files_contents:Sexp.Ast.t Path.Map.t
-    -> f:(String_with_vars.t -> Value.t list)
-    -> expanded
+    -> f:expander
+    -> Partial.t * Sexp.syntax * Path.Set.t
```
Comment (diml) [inline src/ordered_set_lang.mli]: It feels a bit weird to return the syntax here, what about adding a function `val syntax : t -> Sexp.syntax`?

File: src/string_with_vars.ml [hunk 3]
```diff
@@ -179,9 +179,9 @@ module Mode = struct
 end
 
 module Partial = struct
-  type nonrec 'a t =
-    | Expanded of 'a
-    | Unexpanded of t
+  type nonrec ('expanded, 'unexpanded) t =
```
Comment (diml) [inline src/string_with_vars.ml]: Why this change? The original type looked simpler and looking at the use cases for `Partial.t`, it seems that `'unexpanded` is always `String_with_vars.t`

File: src/super_context.ml [hunk 4]
```diff
@@ -95,52 +132,328 @@ let expand_ocaml_config t pform name =
...
-    Ordered_set_lang.Unexpanded.files set ~f in
-  let f = expand_vars t ~mode:Many ~scope ~dir ?bindings in
-  match Path.Set.to_list files with
-  | [] ->
-    let set =
-      Ordered_set_lang.Unexpanded.expand set ~dir
-        ~files_contents:Path.Map.empty ~f
+  let bindings = Option.value ~default:Pform.Map.empty bindings in
+  let (partial, syntax, paths, resolved_forms) =
+    let ((partial, syntax, paths), resolved_forms) =
```
Comment (diml) [inline src/super_context.ml]: This inner let feels extra

File: src/super_context.ml [hunk 5]
```diff
@@ -95,52 +132,328 @@ let expand_ocaml_config t pform name =
...
+      Expander.with_expander t ~dir ~dep_kind:Required
+        ~scope ~targets_written_by_user:(Static [])
+        ~map_exe:(fun x -> x)
```
Comment (diml) [inline src/super_context.ml]: This `map_exe` should be the same as for actions.

File: src/ordered_set_lang.ml [hunk 6]
```diff
@@ -141,6 +147,7 @@ module Make(Key : Key)(Value : Value with type key = Key.t) = struct
   end
 
   module Make(M : Named_values) = struct
+    [@@@warning "-56"] (* XXX necessary for 4.02.3 *)
```
[the remark this answers is not in the document]
Comment (mjambon) [reply src/ordered_set_lang.ml]: FYI I just suggested mnemonic aliases for warning codes on mantis: https://caml.inria.fr/mantis/view.php?id=7823

[the remark this answers is not in the document]
Comment (diml) [reply src/ordered_set_lang.ml]: That would help indeed :)

File: src/string_with_vars.ml (hunk 3 above)
Comment (rgrinberg) [reply src/string_with_vars.ml]: It's because before I wanted the type of `Ordered_set_lang.Unexpanded.expand` t o be:

```ocaml
val expand
  :  t
  -> dir:Path.t
  -> f:(String_with_vars.t -> (Value.t list, Strign_with_vars.t) String_with_partial.Partial.t))
  -> (expanded, Partial.t) String_with_vars.Partial.t
```

where the point was that we could short circuit the `Osl.Partial.t` step if we managed to expand everything. I never managed to get this to work, and I'm not even sure it's possible to do it efficiently after a recent try. So I will undo this change.

File: src/ordered_set_lang.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/ordered_set_lang.ml]: My comment is a bit inaccurate I guess. You will get warning 56 for 4.06.1:

```
File "src/ordered_set_lang.ml", line 166, characters 10-19:
Error (warning 56): this match case is unreachable.
Consider replacing it with a refutation case '<pat> -> .'
```

But we cannot follow the advice of this warning because it wouldn't work in 4.02.3

[the remark this answers is not in the document]
Comment (diml) [reply src/ordered_set_lang.ml]: I see, thanks for the explanation

File: src/ordered_set_lang.ml [hunk 7]
```diff
@@ -212,6 +223,52 @@ let standard =
...
 
+module Partial = struct
+  type ast =
+    ((Loc.t * Value.t list, String_with_vars.t) String_with_vars.Partial.t
+    , Ast.partial
+    ) Ast.t
+
+  type t = ast generic
+
+  let expand t ~dir ~files_contents ~(f : String_with_vars.t -> Value.t list) =
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/ordered_set_lang.ml]: What kind of signature for `eval_partial` are you consider considering? This?

```
val eval_partial
  :  t
  -> parse:(loc:Loc.t -> string -> value)
  -> dir:Path.t
  -> files_contents:Sexp.Ast.t Path.Map.t
  -> f:(String_with_vars.t -> Value.t list)
  -> standard:value list
  -> value list
```

Where `t` would be basically what I have as `Oredered_set_lang.Partial.t` now

I'd like to make things more efficient but this signature is quite busy already, and the stages don't really have well defined scope anymore.

We do have cases where we are expanding the OSL without any variables (for example in `parse_modules`). It doesn't seem so nice to complicate those uses. Also, I'm not entirely sure how we'd reconcile `Value.t`'s with `value`. We could do it via `parse (Value.to_string value)`, but this seems like what we're trying to avoid.

[the remark this answers is not in the document]
Comment (diml) [reply src/ordered_set_lang.ml]: I was thinking to have this in addition to the current evaluation functions. I forgot `value` was abstract, I suppose this function needs to go outside of the functor then. As a result, we shouldn't need the `parse` function, and maybe not the `dir` argument. This should simplify the signature.

[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/ordered_set_lang.ml]: Okay, that's fine. We'll still have this annoying duplication between the eval's, but at least it will localized to the Osl module. I suppose I'll only implement the ordered variant of evaluation here, as the other isn't used with templates.

[the remark this answers is not in the document]
Comment (diml) [reply src/ordered_set_lang.ml]: seems good

[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/ordered_set_lang.ml]: Okay, this is what I've done. As discussed, the comparison is done as a string which necessitates the `~dir` arg after all. It might also be just simpler to return `string list` from `expand`.

**Comment by diml** [comment]
@rgrinberg I'm reading the latest changes, could you add tests for the various cases we discussed on slack?

File: src/super_context.ml [hunk 8]
```diff
@@ -95,52 +132,336 @@ let expand_ocaml_config t pform name =
...
+      ~bindings
+      ~f:(fun f ->
+        let f : Ordered_set_lang.Unexpanded.expander =
+          { f = fun ~mode sw ->
+              String_with_vars.partial_expand ~mode ~dir ~f sw
+          } in
+        Ordered_set_lang.Unexpanded.expand set ~dir ~f)
+  in
+  let f =
+    Expander.Resolved_forms.build resolved_forms >>^ fun dynamic_expansions ->
```
Comment (diml) [inline src/super_context.ml]: This is supposed to be called only once. Additionally, `resolved_forms` is mutable, so this should only be called once all the mutations have been done.

Comment (rgrinberg) [reply src/super_context.ml]: It is called only once. Also, all the mutations occurred in the call to `with_expander` above it.

**Comment by rgrinberg** [comment]
As discussed in slack the approach in this PR is fatally flawed. The main problem is that the dependency calculation in OSL should take after the expansion. This is because variables in negative position (rhs of `\`) should not introduce dependencies. For example:

```
(foo \ %{bin:bar})
```

Should not add `%{bin:bar}` as a dependency.
